### PR TITLE
Renaming MasterSlave -> IntraComm, Master -> Primary, Slave -> Secondary

### PR DESCRIPTION
**Pull request by IshaanDesai**
## Main changes of this PR

The following renaming is done:
* `MasterSlave -> IntraComm`
* `Master -> Primary`
* `Slave -> Secondary`
* `MASTER -> PRIMARY`
* `SLAVES -> SECONDARIES`

Several files having `MasterSlave` or `Master` in their names have also been renamed. The `<master:>` tag in the XML configuration has been changed to `<primary:>` tag. If a user still uses the `master` tag, a deprecated warning will be thrown.

## Motivation and additional information

Its about time these changes happen :fist_raised: 

## Author's checklist

* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I ran `make format` to ensure everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.10.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

File: src/m2n/BoundM2N.cpp [hunk 1]
```diff
@@ -21,36 +21,36 @@ void BoundM2N::prepareEstablishment()
   m2n->prepareEstablishment(localName, remoteName);
 }
 
-void BoundM2N::connectMasters()
+void BoundM2N::connectPrimarys()
```
Comment (fsimonis) [inline src/m2n/BoundM2N.cpp:24]:
```suggestion
void BoundM2N::connectPrimaries()
```

File: src/precice/config/ParticipantConfiguration.cpp [hunk 2]
```diff
@@ -164,8 +164,8 @@ ParticipantConfiguration::ParticipantConfiguration(
                                "processors. Both result in the same distribution (if the safety factor is sufficiently large). "
                                "\"on-master\" is not supported if you use two-level initialization. "
```
Comment (fsimonis) [inline src/precice/config/ParticipantConfiguration.cpp:165]: These filters should be updated too.

File: src/precice/config/ParticipantConfiguration.cpp [hunk 3]
```diff
@@ -164,8 +164,8 @@ ParticipantConfiguration::ParticipantConfiguration(
...
                                "For very asymmetric cases, the filter can also be switched off completely (\"no-filter\").")
-                           .setOptions({VALUE_FILTER_ON_MASTER, VALUE_FILTER_ON_SLAVES, VALUE_NO_FILTER})
-                           .setDefaultValue(VALUE_FILTER_ON_SLAVES);
+                           .setOptions({VALUE_FILTER_ON_PRIMARY, VALUE_FILTER_ON_SECONDARIES, VALUE_NO_FILTER})
```
Comment (fsimonis) [inline src/precice/config/ParticipantConfiguration.cpp:167]: These filters should be backwards compatible too. As an example both `on-master` and `on-primary` should work.

File: src/precice/config/ParticipantConfiguration.cpp [hunk 4]
```diff
@@ -191,62 +191,62 @@ ParticipantConfiguration::ParticipantConfiguration(
   std::list<XMLTag>  masterTags;
   XMLTag::Occurrence masterOcc = XMLTag::OCCUR_NOT_OR_ONCE;
   {
-    XMLTag tagMaster(*this, "sockets", masterOcc, TAG_MASTER);
+    XMLTag tagPrimary(*this, "sockets", masterOcc, TAG_PRIMARY);
```
Comment (fsimonis) [inline src/precice/config/ParticipantConfiguration.cpp:194]: This should be called something along the lines of intra or intracomm.

File: src/precice/config/ParticipantConfiguration.cpp [hunk 5]
```diff
@@ -191,62 +191,62 @@ ParticipantConfiguration::ParticipantConfiguration(
...
 
-    masterTags.push_back(tagMaster);
+    masterTags.push_back(tagPrimary);
   }
   {
-    XMLTag tagMaster(*this, "mpi", masterOcc, TAG_MASTER);
+    XMLTag tagPrimary(*this, "mpi", masterOcc, TAG_PRIMARY);
     doc = "A solver in parallel needs a communication between its ranks. ";
     doc += "By default, the participant's MPI_COM_WORLD is reused.";
     doc += "Use this tag to use MPI with separated communication spaces instead instead.";
```
Comment (fsimonis) [inline src/precice/config/ParticipantConfiguration.cpp:228]:
```suggestion
    doc += "Use this tag to use MPI with separated communication spaces instead.";
```

File: src/precice/config/ParticipantConfiguration.cpp [hunk 6]
```diff
@@ -374,10 +377,10 @@ ParticipantConfiguration::getParticipants() const
 
 partition::ReceivedPartition::GeometricFilter ParticipantConfiguration::getGeoFilter(const std::string &geoFilter) const
 {
-  if (geoFilter == VALUE_FILTER_ON_MASTER) {
-    return partition::ReceivedPartition::GeometricFilter::ON_MASTER;
-  } else if (geoFilter == VALUE_FILTER_ON_SLAVES) {
-    return partition::ReceivedPartition::GeometricFilter::ON_SLAVES;
+  if (geoFilter == VALUE_FILTER_ON_PRIMARY) {
```
Comment (fsimonis) [inline src/precice/config/ParticipantConfiguration.cpp:380]: As mentioned above. These should be backwards compatible.

File: src/precice/impl/SolverInterfaceImpl.cpp [hunk 7]
```diff
@@ -238,8 +238,8 @@ void SolverInterfaceImpl::configure(
   utils::EventRegistry::instance().initialize("precice-" + _accessorName, "", utils::Parallel::current()->comm);
 
   PRECICE_DEBUG("Initialize master-slave communication");
-  if (utils::MasterSlave::isParallel()) {
-    initializeMasterSlaveCommunication();
+  if (utils::IntraComm::isParallel()) {
+    initializeIntraCommCommunication();
```
Comment (fsimonis) [inline src/precice/impl/SolverInterfaceImpl.cpp:242]: *cough*

**Comment by fsimonis** [comment]
How is this PR related to #1253?

**Comment by IshaanDesai** [comment]
> How is this PR related to #1253?

This PR is not directly related to #1253 as the changes in both are strictly independent of each other. The decoupling is done on purpose to make the reviewing manageable.

File: src/precice/config/ParticipantConfiguration.cpp (hunk 2 above)
Comment (IshaanDesai) [reply src/precice/config/ParticipantConfiguration.cpp:165]: This filter is updated in https://github.com/precice/precice/pull/1253

File: src/acceleration/impl/ParallelMatrixOperations.hpp [hunk 8]
```diff
@@ -280,28 +280,28 @@ class ParallelMatrixOperations {
...
+    if (utils::IntraComm::isPrimary()) {
       // distribute blocks of summarizedBlocks (result of multiplication) to corresponding slaves
       result = summarizedBlocks.block(0, 0, offsets[1], r);
 
-      for (Rank rankSlave : utils::MasterSlave::allSlaves()) {
-        int off       = offsets[rankSlave];
-        int send_rows = offsets[rankSlave + 1] - offsets[rankSlave];
+      for (Rank rankSecondary : utils::IntraComm::allSecondaries()) {
+        int off       = offsets[rankSecondary];
+        int send_rows = offsets[rankSecondary + 1] - offsets[rankSecondary];
```
Comment (DavidSCN) [inline src/acceleration/impl/ParallelMatrixOperations.hpp:298]: I'm not sure how many times this occurs, but we should maybe rename it to `secondaryRank`

```suggestion
      for (Rank secondaryRank : utils::IntraComm::allSecondaries()) {
        int off       = offsets[secondaryRank];
        int send_rows = offsets[secondaryRank + 1] - offsets[secondaryRank];
```

File: src/acceleration/impl/ParallelMatrixOperations.hpp [hunk 9]
```diff
@@ -280,28 +280,28 @@ class ParallelMatrixOperations {
...
 
         if (summarizedBlocks.block(off, 0, send_rows, r).size() > 0) {
           // necessary to save the matrix-block that is to be sent in a temporary matrix-object
           // otherwise, the send routine walks over the bounds of the block (matrix structure is still from the entire matrix)
           Eigen::MatrixXd sendBlock = summarizedBlocks.block(off, 0, send_rows, r);
-          utils::MasterSlave::getCommunication()->send(sendBlock, rankSlave);
+          utils::IntraComm::getCommunication()->send(sendBlock, rankSecondary);
```
Comment (DavidSCN) [inline src/acceleration/impl/ParallelMatrixOperations.hpp:304]:
```suggestion
          utils::IntraComm::getCommunication()->send(sendBlock, secondaryRank);
```

File: src/acceleration/impl/QRFactorization.cpp [hunk 10]
```diff
@@ -359,14 +359,14 @@ int QRFactorization::orthogonalize(
...
+    rho1 = utils::IntraComm::l2norm(v); // distributed l2norm
 
     // t = norm of r_(:,j) with j = colNum-1
-    double norm_coefficients = utils::MasterSlave::l2norm(s); // distributed l2norm
+    double norm_coefficients = utils::IntraComm::l2norm(s); // distributed l2norm
     k++;
 
     // treat the special case m=n
-    // Attention (Master-Slave): Here, we need to compare the global _rows with colNum and NOT the local
+    // Attention (Primary-Secondary): Here, we need to compare the global _rows with colNum and NOT the local
```
Comment (DavidSCN) [inline src/acceleration/impl/QRFactorization.cpp:369]: Isn't this referring to `IntraComm`?

File: src/acceleration/impl/QRFactorization.cpp [hunk 11]
```diff
@@ -483,14 +483,14 @@ int QRFactorization::orthogonalize_stable(
...
+    rho1 = utils::IntraComm::l2norm(v); // distributed l2norm
 
     // t = norm of r_(:,j) with j = colNum-1
-    t = utils::MasterSlave::l2norm(s); // distributed l2norm
+    t = utils::IntraComm::l2norm(s); // distributed l2norm
     k++;
 
     // treat the special case m=n
-    // Attention (Master-Slave): Here, we need to compare the global _rows with colNum and NOT the local
+    // Attention (Primary-Secondary): Here, we need to compare the global _rows with colNum and NOT the local
```
Comment (DavidSCN) [inline src/acceleration/impl/QRFactorization.cpp:493]: Same as above

File: src/acceleration/impl/QRFactorization.cpp [hunk 12]
```diff
@@ -560,18 +560,18 @@ int QRFactorization::orthogonalize_stable(
...
           global_uk = u(k);
-          for (Rank rankSlave : utils::MasterSlave::allSlaves()) {
-            utils::MasterSlave::getCommunication()->receive(local_k, rankSlave);
-            utils::MasterSlave::getCommunication()->receive(local_uk, rankSlave);
+          for (Rank rankSecondary : utils::IntraComm::allSecondaries()) {
+            utils::IntraComm::getCommunication()->receive(local_k, rankSecondary);
+            utils::IntraComm::getCommunication()->receive(local_uk, rankSecondary);
             if (local_uk < global_uk) {
-              rank      = rankSlave;
+              rank      = rankSecondary;
```
Comment (DavidSCN) [inline src/acceleration/impl/QRFactorization.cpp:574]: `secondaryRank`?

File: src/com/CommunicateBoundingBox.hpp [hunk 13]
```diff
@@ -48,7 +48,7 @@ class CommunicateBoundingBox {
   void broadcastSendBoundingBoxMap(
       mesh::Mesh::BoundingBoxMap &bbm);
 
-  /// Slaves call this method to receive the set of bounding boxes sent by the master.
+  /// Secondaries call this method to receive the set of bounding boxes sent by the master.
```
Comment (DavidSCN) [inline src/com/CommunicateBoundingBox.hpp:51]:
```suggestion
  /// Secondary ranks call this method to receive the set of bounding boxes sent by the master.
```

File: src/com/Communication.cpp [hunk 14]
```diff
@@ -57,12 +57,12 @@ void Communication::reduceSum(precice::span<double const> itemsToSend, precice::
   }
 }
 
-void Communication::reduceSum(precice::span<double const> itemsToSend, precice::span<double> itemsToReceive, Rank rankMaster)
+void Communication::reduceSum(precice::span<double const> itemsToSend, precice::span<double> itemsToReceive, Rank rankPrimary)
```
Comment (DavidSCN) [inline src/com/Communication.cpp:60]: ?
```suggestion
void Communication::reduceSum(precice::span<double const> itemsToSend, precice::span<double> itemsToReceive, Rank primaryRank)
```

File: src/com/Communication.hpp [hunk 15]
```diff
@@ -168,18 +168,18 @@ class Communication {
                                          std::set<int> const &acceptorRanks,
                                          int                  requesterRank) = 0;
 
-  /** Establishes the Master-Slave connection.
+  /** Establishes the Primary-Secondary connection.
```
Comment (DavidSCN) [inline src/com/Communication.hpp:171]: ?
```suggestion
  /** Establishes the Intra participant communication connection.
```

File: src/precice/config/ParticipantConfiguration.cpp [hunk 16]
```diff
@@ -374,10 +379,16 @@ ParticipantConfiguration::getParticipants() const
 
 partition::ReceivedPartition::GeometricFilter ParticipantConfiguration::getGeoFilter(const std::string &geoFilter) const
 {
-  if (geoFilter == VALUE_FILTER_ON_MASTER) {
-    return partition::ReceivedPartition::GeometricFilter::ON_MASTER;
-  } else if (geoFilter == VALUE_FILTER_ON_SLAVES) {
-    return partition::ReceivedPartition::GeometricFilter::ON_SLAVES;
+  if (geoFilter == VALUE_FILTER_ON_MASTER || geoFilter == VALUE_FILTER_ON_PRIMARY) {
+    if (geoFilter == VALUE_FILTER_ON_MASTER) {
+      PRECICE_WARN("<on-master> tag is deprecated and will be removed in v3.0.0! Please use <on-primary> tag instead");
```
Comment (DavidSCN) [inline src/precice/config/ParticipantConfiguration.cpp:384]:
```suggestion
      PRECICE_WARN("<on-master> tag is deprecated and will be removed in v3.0.0. Please use <on-primary> tag instead");
```
We decided to not use exclamation marks in error messages.

File: src/precice/config/ParticipantConfiguration.cpp [hunk 17]
```diff
@@ -374,10 +379,16 @@ ParticipantConfiguration::getParticipants() const
...
+    }
+    return partition::ReceivedPartition::GeometricFilter::ON_PRIMARY;
+  } else if (geoFilter == VALUE_FILTER_ON_SLAVES || geoFilter == VALUE_FILTER_ON_SECONDARIES) {
+    if (geoFilter == VALUE_FILTER_ON_SLAVES) {
+      PRECICE_WARN("<on-slaves> tag is deprecated and will be removed in v3.0.0! Please use <on-secondaries> tag instead");
```
Comment (DavidSCN) [inline src/precice/config/ParticipantConfiguration.cpp:389]:
```suggestion
      PRECICE_WARN("<on-slaves> tag is deprecated and will be removed in v3.0.0. Please use <on-secondaries> tag instead");
```

File: src/utils/IntraComm.hpp [hunk 18]
```diff
@@ -15,8 +15,8 @@ class Logger;
 
 namespace utils {
 
-/// Utility class for managing Master-Slave operations.
-class MasterSlave {
+/// Utility class for managing Primary-Secondary operations.
```
Comment (DavidSCN) [inline src/utils/IntraComm.hpp:18]: ?
```suggestion
/// Utility class for managing intra-participant communication operations.
```

**Review by DavidSCN — COMMENTED** [review]
Only minor comments. I'm not sure about the variable names such as `rankSecondary` etc. For me `secondaryRank` would sound more natural, but I will leave it up to you.

**Comment by IshaanDesai** [comment]
This PR has diverged and is hence discontinued. The changes are now pursued through https://github.com/precice/precice/pull/1274

**Pull request closed without merging**